packages/core/src/browser/view-container.ts

```diff
@@ -586,36 +704,115 @@ export class ViewContainer extends BaseWidget implements StatefulWidget, Applica
-                        this.moveBefore(moveId, part.id);
+                    if (ViewContainer.isDNDSupported(event)) {
+                        const moveId = dataTransfer.getData(ViewContainer.DataTransfers.DND);
+                        const containerDescription = dataTransfer.getData(ViewContainer.DataTransfers.FROM_CONTAINER_DESCRIPTION);
+                        if (moveId && moveId !== part.id) {
+                            this.movePart(containerDescription, moveId, part.id);
+                        }
+                        this.toDisposeOnDragEnd.dispose();
+                    } else if (ViewContainer.isInternalDND(event)) {
+                        const moveId = dataTransfer.getData(ViewContainer.DataTransfers.INTERNAL_DND);
+                        if (moveId && moveId !== part.id) {
+                            this.moveBefore(moveId, part.id);
+                        }
                     }
                     unstyle(event);
                 }
             }, false)
         );
     }
 
+    toDisposeOnDragEnd = new DisposableCollection();
+    protected onDragEnterOrLeave(event: DragEvent): void {
+        if (!ViewContainer.isDNDSupported(event)) {
+            return;
+        }
+        this.toDisposeOnDragEnd.dispose();
+    }
+
+    protected onDragOver(event: DragEvent): void {
+        if (!ViewContainer.isDNDSupported(event) || this.disableDNDBetweenContainers) {
+            if (event.dataTransfer) {
+                event.dataTransfer.dropEffect = 'none';
+            }
+            return;
+        }
+        const containerDockPanel = this.isDebugContainer() ? this.parent!.parent!.parent : this.parent;
+        if (containerDockPanel instanceof DockPanel) {
+            const { overlay } = containerDockPanel;
+            overlay.show({ top: 0, bottom: 0, right: 0, left: 0 });
+
+            this.toDisposeOnDragEnd.push(Disposable.create(() =>
+                overlay.hide(100)
+            ));
+        }
+    }
+
+    protected onDrop(event: DragEvent): void {
+        const { dataTransfer } = event;
+        if (!dataTransfer || !ViewContainer.isDNDSupported(event) || this.disableDNDBetweenContainers) {
+            return;
+        }
+        const moveId = dataTransfer.getData(ViewContainer.DataTransfers.DND);
+        const containerDescription = dataTransfer.getData(ViewContainer.DataTransfers.FROM_CONTAINER_DESCRIPTION);
+        if (moveId) {
+            this.movePart(containerDescription, moveId);
+        }
+        this.toDisposeOnDragEnd.dispose();
+    }
+
+    protected isDebugContainer(): boolean {
+        return !!this.parent && !!this.parent.parent && this.isDebugWidget(this.parent.parent);
+    }
+
+    protected isDebugWidget(widget: Widget): boolean {
+        return 'sessionWidget' in widget;
+    }
+
+    protected getDebugWidgetId(): string | undefined {
+        if (this.isDebugContainer()) {
+            return this.parent!.parent!.id;
+        }
+    }
```

Comment:
If possible, we should avoid anything debug related in `core`. Perhaps define an interface here that the debug widget could implement that would get around these problems?

Reply:
@colin-grant-work 

1. should I create an interface with the name `debug` (referring only the name)?
2. Which file is the right place to put the interface in?

Reply:
What I have in mind is something like
In view-container.ts:
```typescript
export interface ViewContainerContainer extends Widget {
    getContainedViewContainer(): ViewContainer
}

export namespace ViewContainerContianer {
    export const is = (candidate: Widget): candidate is ViewContainerContainer => {
       const maybeViewContainerContainer = candidate as ViewContainerContainer;
       return 'getContainedViewContainer' in maybeViewContainerContainer && typeof maybeViewContainerContainer.getContainedViewContainer === 'function';
};
```

Then in `debug-widget.ts`
```typescript
export class DebugWidget extends BaseWidget implements StatefulWidget, ApplicationShell.TrackableWidgetProvider, ViewContainerContainer {

getContainedViewContainer(): ViewContainer {
    return this.sessioWidget.viewContainer; 
    // this is protected, so it would either have to made public, or SessionWidget would implement ViewContainerContainer and you could return this.sessionWidget.getContainedViewContainer();
}
```

That way, the `debug` package would know about `core`, but `core` wouldn't have to know anything about `debug` - not even the name.